### PR TITLE
fix: Remove duplicate vertices from well path. May cause crash

### PR DESCRIPTION
**Pull request by nilscb**

### CODE REVIEW

**Review by hkfb — CHANGES_REQUESTED** [review]
Can we do this in `initializeState` instead, so that duplicates are also removed when not interpolating?

Would it be useful to get a warning if duplicates occur?

Also, a unit test would be good for this.

File: typescript/packages/subsurface-viewer/src/layers/wells/utils/spline.test.ts
Comment (hkfb) [inline typescript/packages/subsurface-viewer/src/layers/wells/utils/spline.test.ts:1]: Would be good to capture also corner cases, like eg. an empty set of coordinates

File: typescript/packages/subsurface-viewer/src/layers/wells/utils/spline.ts [hunk 1]
```diff
@@ -7,6 +7,72 @@ import { cloneDeep, range } from "lodash";
...
+        }
+
+        let coords = lineString.coordinates as Position3D[];
+
+        const nOrig = coords.length;
+        [coords, mds[0]] = removeConsecutiveDuplicates(coords, mds[0]);
+
+        const n = coords.length;
+        if (n != nOrig) {
+            console.warn("Well no ", well_no, " contains duplicates.");
```
Comment (hkfb) [inline typescript/packages/subsurface-viewer/src/layers/wells/utils/spline.ts:57]: "Well no" is confusing. Better to spell it out.

File: typescript/packages/subsurface-viewer/src/layers/wells/utils/spline.ts [hunk 2]
```diff
@@ -246,14 +315,29 @@ export function coarsenWells(data_in: FeatureCollection): FeatureCollection {
...
-
-        const coordsSimplified = simplify(lineString, options);
-        lineString.coordinates = coordsSimplified.coordinates as Position3D[];
+        const n = lineString.coordinates.length;
+        const isHorizontalWell =
+            lineString.coordinates[0][0] == lineString.coordinates[n - 1][0] &&
+            lineString.coordinates[0][1] == lineString.coordinates[n - 1][1];
+
+        if (isHorizontalWell) {
+            // The simplify algorithm below did not work on horizontal wells hence in this case we only use first and last point.
```
Comment (hkfb) [inline typescript/packages/subsurface-viewer/src/layers/wells/utils/spline.ts:324]: Do you mean "vertical"?

File: typescript/packages/subsurface-viewer/src/layers/wells/utils/spline.ts [hunk 3]
```diff
@@ -246,14 +315,29 @@ export function coarsenWells(data_in: FeatureCollection): FeatureCollection {
...
-            highQuality: false,
-            mutate: false,
-        };
-
-        const coordsSimplified = simplify(lineString, options);
-        lineString.coordinates = coordsSimplified.coordinates as Position3D[];
+        const n = lineString.coordinates.length;
+        const isHorizontalWell =
+            lineString.coordinates[0][0] == lineString.coordinates[n - 1][0] &&
+            lineString.coordinates[0][1] == lineString.coordinates[n - 1][1];
```
Comment (hkfb) [inline typescript/packages/subsurface-viewer/src/layers/wells/utils/spline.ts:321]: This doesn't look like a proper check for a horizontal or vertical well

File: typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.stories.tsx [hunk 4]
```diff
@@ -437,6 +437,40 @@ Wells3d.parameters = {
     },
 };
 
+export const OneWell = Template.bind({});
+OneWell.args = {
+    id: "well_176",
+    bounds: [756125, 166200, 757327, 167938],
+    layers: [
+        {
+            "@@type": "WellsLayer",
```
Comment (hkfb) [inline typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.stories.tsx:446]: Better to use the class directly in order to get proper type checking.

**Review by hkfb — CHANGES_REQUESTED** [review]

File: typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.stories.tsx [hunk 5]
```diff
@@ -437,6 +437,40 @@ Wells3d.parameters = {
...
+            data: "./well_176.json",
```
Comment (hkfb) [inline typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.stories.tsx:447]: Do we need more well data examples?

File: typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.stories.tsx [hunk 6]
```diff
@@ -437,6 +437,40 @@ Wells3d.parameters = {
     },
 };
 
+export const OneWell = Template.bind({});
```
Comment (hkfb) [inline typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.stories.tsx:440]: This story name is not very descriptive.

**Review by hkfb — CHANGES_REQUESTED** [review]

File: typescript/packages/subsurface-viewer/src/layers/wells/utils/spline.ts (hunk 3 above)
Comment (nilscb) [reply typescript/packages/subsurface-viewer/src/layers/wells/utils/spline.ts:321]: I think this is sufficient in 99.9999% of the cases. But I could of course di a more expensive test and use all points in the well path.

File: example-data/one_well_duplicate_vertexes.json
Comment (hkfb) [inline example-data/one_well_duplicate_vertexes.json:1]: Maybe nitpicking now, but normally it's called "vertices"

Comment (nilscb) [reply example-data/one_well_duplicate_vertexes.json:1]: Good point! Spelling is important! (I learned something new :-)

File: typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.stories.tsx [hunk 7]
```diff
@@ -437,6 +437,70 @@ Wells3d.parameters = {
...
+WellWithDuplicates.args = {
+    id: "well_176",
+    bounds: [756125, 166200, 757327, 167938],
+    layers: [
+        new WellsLayer({
+            data: "./one_well_duplicate_vertices.json",
+            refine: true,
+            simplifiedRendering: true,
+        }),
+    ],
```
Comment (hkfb) [inline typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.stories.tsx:481]: I think it would be better to create some simple synthetic well trajectories in the stories themselves instead of adding more well files.

**Comment by hkfb** [comment]
The well is invisible until I zoom out: 
image
image

**Review by hkfb — CHANGES_REQUESTED** [review]

File: typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.ts [hunk 8]
```diff
@@ -248,6 +249,8 @@ export default class WellsLayer extends CompositeLayer<WellsLayerProps> {
                 data = invertPath(data);
             }
 
+            removeDuplicates(data);
```
[the remark this answers is not in the document]
Comment (nilscb) [reply typescript/packages/subsurface-viewer/src/layers/wells/wellsLayer.ts:252]: In the unit test the function "removeConsecutiveDuplicates" is tested. But I changed "removeDuplicates" so that it may only be used in one way.
